lib/elixir/lib/calendar.ex

```diff
@@ -1867,11 +2198,137 @@ defmodule DateTime do
       :gt
   """
   @spec compare(DateTime.t, DateTime.t) :: :lt | :eq | :gt
-  def compare(%DateTime{} = datetime1, %DateTime{} = datetime2) do
-    case {to_unix(datetime1, :microsecond), to_unix(datetime2, :microsecond)} do
+  def compare(%DateTime{utc_offset: utc_offset1, std_offset: std_offset1} = datetime1, %DateTime{utc_offset: utc_offset2, std_offset: std_offset2} = datetime2) do
+    {days1, {parts1, ppd1}} =
+      to_rata_die(datetime1)
+      |> apply_tz_offset(utc_offset1)
+      |> apply_tz_offset(std_offset1)
+    {days2, {parts2, ppd2}} =
+      to_rata_die(datetime2)
+      |> apply_tz_offset(utc_offset2)
+      |> apply_tz_offset(std_offset2)
+
+    # Ensure fraction tuples have same denominator.
+    rata_die1 = {days1, parts1 * ppd2}
+    rata_die2 = {days2, parts2 * ppd1}
+
+    case {rata_die1, rata_die2}  do
       {first, second} when first > second -> :gt
       {first, second} when first < second -> :lt
       _ -> :eq
     end
   end
+
+  @doc """
+  Returns the difference between two `DateTime` structs,
+  in the Calendar.rata_die format: {days, day_fraction}
+
+  ## Examples
+
+      iex> dt1 = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                 hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                 utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> dt2 = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "CET",
+      ...>                 hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                 utc_offset: 3600, std_offset: 0, time_zone: "Europe/Warsaw"}
+      iex> DateTime.diff(dt1, dt2)
+      {0, {5, 24}}
+  """
+  @spec diff(DateTime.t, DateTime.t) :: Calendar.rata_die
+  def diff(%DateTime{utc_offset: utc_offset1, std_offset: std_offset1} = datetime1, %DateTime{utc_offset: utc_offset2, std_offset: std_offset2} = datetime2) do
```

Comment:
It should be implemented like NaiveDateTime diff and return a given unit in Calendar.ISO.